File: backend/src/main/java/com/styp/cenate/service/dengue/DengueService.java
package com.styp.cenate.service.dengue;

import com.styp.cenate.dto.dengue.DengueImportResultDTO;
import com.styp.cenate.model.bolsas.SolicitudBolsa;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

/**
 * Interfaz de servicio para gestión del módulo Dengue
 *
 * Define operaciones para:
 * - Importación masiva de Excel (6,548 registros)
 * - Listado y búsqueda de casos Dengue
 * - Integración con:
 *   1. CIE-10 (validación de dx_main)
 *   2. IPRESS (lookup de IPRESS + Red)
 *   3. Asegurados (búsqueda/creación de pacientes)
 *   4. dim_solicitud_bolsa (almacenamiento final)
 *
 * @version 1.0.0
 * @since 2026-01-29
 */
public interface DengueService {

    /**
     * Carga un archivo Excel con casos de dengue
     *
     * Procesa 6,548 registros aplicando:
     * 1. Normalización de DNI (8 dígitos)
     * 2. Validación CIE-10 (A97.0/A97.1/A97.2)
     * 3. Lookup de IPRESS por cenasicod
     * 4. Búsqueda/creación en asegurados
     * 5. Deduplicación (DNI + fecha_atencion)
     *
     * @param archivo MultipartFile con Excel (.xlsx)
     * @param usuarioId ID del usuario que carga el archivo (para auditoría)
     * @return DengueImportResultDTO con estadísticas (insertados, actualizados, errores)
     * @throws IOException Si hay error leyendo el archivo
     * @throws RuntimeException Si hay error procesando registros
     */
    DengueImportResultDTO cargarExcelDengue(MultipartFile archivo, Long usuarioId);

    /**
     * Lista todos los casos de dengue con paginación
     *
     * @param pageable Información de paginación (page, size, sort)
     * @return Page con casos de dengue ordenados por fecha_solicitud DESC
     */
    Page<SolicitudBolsa> listarCasosDengue(Pageable pageable);

    /**
     * Busca casos de dengue con filtros opcionales
     *
     * @param dni DNI del paciente (opcional, búsqueda parcial)
     * @param dxMain Código CIE-10 (opcional, búsqueda exacta: A97.0, A97.1, A97.2)
     * @param pageable Información de paginación
     * @return Page con casos que coinciden con los filtros
     */
    Page<SolicitudBolsa> buscarCasosDengue(String dni, String dxMain, Pageable pageable);

    /**
     * Obtiene un caso específico por ID
     *
     * @param idSolicitud ID de la solicitud de bolsa
     * @return SolicitudBolsa con el caso dengue
     * @throws RuntimeException Si no existe el registro
     */
    SolicitudBolsa obtenerCasoDengue(Long idSolicitud);

}

File: frontend/src/services/dengueService.js
import apiClient from '../lib/apiClient';

/**
 * 🦟 Servicio de API - Módulo Dengue
 * v1.0.0 - Gestión de casos dengue e importación desde Excel
 *
 * Endpoints:
 * - POST /api/dengue/cargar-excel - Cargar Excel con casos
 * - GET /api/dengue/casos - Listar casos con paginación
 * - GET /api/dengue/buscar - Buscar con filtros
 * - GET /api/dengue/estadisticas - Estadísticas
 */

const API_BASE_URL = '/dengue';

/**
 * Carga un archivo Excel con casos dengue
 * @param {File} archivo - Archivo Excel (.xlsx)
 * @param {number} usuarioId - ID del usuario que realiza la carga
 * @returns {Promise<Object>} - DengueImportResultDTO con resultados
 */
export const cargarExcelDengue = async (archivo, usuarioId) => {
  try {
    const formData = new FormData();
    formData.append('archivo', archivo);
    formData.append('usuarioId', usuarioId);

    const response = await apiClient.uploadFile(
      `${API_BASE_URL}/cargar-excel`,
      formData
    );
    return response;
  } catch (error) {
    console.error('Error cargando Excel dengue:', error);
    throw error;
  }
};

/**
 * Lista todos los casos dengue con paginación
 * @param {number} page - Número de página (default: 0)
 * @param {number} size - Tamaño de página (default: 30)
 * @param {string} sortBy - Campo de ordenamiento (default: fechaSolicitud)
 * @param {string} sortDirection - ASC o DESC (default: DESC)
 * @returns {Promise<Object>} - Page con casos dengue
 */
export const listarCasosDengue = async (
  page = 0,
  size = 30,
  sortBy = 'fechaSolicitud',
  sortDirection = 'DESC'
) => {
  try {
    const response = await apiClient.get(`${API_BASE_URL}/casos`, {
      params: {
        page,
        size,
        sortBy,
        sortDirection,
      },
    });
    return response;
  } catch (error) {
    console.error('Error listando casos dengue:', error);
    throw error;
  }
};

/**
 * Busca casos dengue con filtros opcionales
 * @param {Object} filtros - Objeto con filtros
 * @param {string} filtros.dni - DNI del paciente (opcional)
 * @param {string} filtros.dxMain - Código CIE-10 (opcional)
 * @param {number} filtros.page - Número de página (default: 0)
 * @param {number} filtros.size - Tamaño de página (default: 30)
 * @param {string} filtros.sortBy - Campo de ordenamiento (default: fechaSolicitud)
 * @param {string} filtros.sortDirection - ASC o DESC (default: DESC)
 * @returns {Promise<Object>} - Page con casos filtrados
 */
export const buscarCasosDengue = async (filtros = {}) => {
  try {
    const {
      dni = '',
      dxMain = '',
      page = 0,
      size = 30,
      sortBy = 'fechaSolicitud',
      sortDirection = 'DESC',
    } = filtros;

    const response = await apiClient.get(`${API_BASE_URL}/buscar`, {
      params: {
        dni: dni || undefined,
        dxMain: dxMain || undefined,
        page,
        size,
        sortBy,
        sortDirection,
      },
    });
    return response;
  } catch (error) {
    console.error('Error buscando casos dengue:', error);
    throw error;
  }
};

/**
 * Obtiene estadísticas de los casos dengue
 * @returns {Promise<Object>} - Objeto con estadísticas
 */
export const obtenerEstadisticasDengue = async () => {
  try {
    const response = await apiClient.get(`${API_BASE_URL}/estadisticas`);
    return response;
  } catch (error) {
    console.error('Error obteniendo estadísticas dengue:', error);
    throw error;
  }
};

/**
 * Hook helper para normalizar errores de API
 */
export const formatearErrorDengue = (error) => {
  if (error.response?.data?.error) {
    return error.response.data.error;
  }
  if (error.response?.data?.detalles) {
    return error.response.data.detalles;
  }
  return error.message || 'Error desconocido';
};

export default {
  cargarExcelDengue,
  listarCasosDengue,
  buscarCasosDengue,
  obtenerEstadisticasDengue,
  formatearErrorDengue,
};
